README.md
# diffutil_sliverlist

[![Pub Package](https://img.shields.io/pub/v/diffutil_sliverlist.svg)](https://pub.dartlang.org/packages/diffutil_sliverlist)
[![Build Status](https://github.com/knaeckeKami/diffutil_sliverlist/workflows/Build/badge.svg)](https://github.com/knaeckeKami/diffutil_sliverlist/actions)
[![codecov](https://codecov.io/gh/knaeckeKami/diffutil_sliverlist/branch/master/graph/badge.svg)](https://codecov.io/gh/knaeckeKami/diffutil_sliverlist)


A SliverList that implicitly animates changes.

It supports two use cases:


## Animating changes from a list of widgets with unique keys

When you have a list of Widgets and can give each Widget a unique key, you can use `DiffUtilSliverList.fromKeyedWidgetList`.

See the example code:

```dart
class _ExpandableListsState extends State<ExpandableLists> {
  bool expanded = false;

  @override
  Widget build(BuildContext context) {
    return DiffUtilSliverList.fromKeyedWidgetList(
      children: [
        ListTile(
          key: Key("1"),
          title: Text("first"),
          trailing: Icon(Icons.chevron_right),
        ),
        ListTile(
          key: Key("2"),
          title: Text("second"),
          trailing: Icon(Icons.chevron_right),
        ),
        if (this.expanded)
          for (int i = 3; i < 6; i++)
            ListTile(
              key: Key(i.toString()),
              title: Text("index: $i"),
              trailing: Icon(Icons.chevron_right),
            ),
        ListTile(
          key: Key("expand_collapse"),
          onTap: () => setState(() {
            expaned = !expaned;
          }),
          title: Text(expanded ? "collapse" : "expand", style: TextStyle(fontWeight: FontWeight.bold),),
          trailing: Icon(expanded ? Icons.expand_less : Icons.expand_more),
        )
      ],
      insertAnimationBuilder: (context, animation, child) => FadeTransition(
        opacity: animation,
        child: child,
      ),
      removeAnimationBuilder: (context, animation, child) => FadeTransition(
        opacity: animation,
        child: SizeTransition(
          sizeFactor: animation,
          axisAlignment: 0,
          child: child,
        ),
      ),
    );
  }
}
```

And the result:

![](https://media.giphy.com/media/UrKN0Se7CCBwTBP01V/giphy.gif)


## Building Widgets from a list of data objects that implement ==

Example:

```dart

Widget build(BuildContext context) {
    return CustomScrollView(
             slivers: [
               DiffUtilSliverList<int>(
                 items: list,
                 builder: (context, item) => 
                   Container(
                     color: colors[item % colors.length],
                     height: 48,
                     width: double.infinity,
                 ),
                 insertAnimationBuilder: (context, animation, child) =>
                     FadeTransition(
                       opacity: animation,
                       child: child,
                 ),
                 removeAnimationBuilder: (context, animation, child) =>
                     SizeTransition(
                       sizeFactor: animation,
                       child: child,
                 ),
                 removeAnimationDuration: const Duration(milliseconds: 3000),
                 insertAnimationDuration: const Duration(milliseconds: 1200),
             ),
           ],
         );
}

```

If `list` changes, the list will automatically animate new/removed items:

![](https://media.giphy.com/media/LRgWnoPvRPW5WEeJYq/giphy.gif)

If the items don't implement `==` correctly, you can pass your own `equalityChecker`.



CHANGELOG.md
## 0.6.0-beta.1

- allow mutations to the passed list by copying the list


## 0.5.1

Relax version constraint on dependeny to diffutil_dart

## 0.5.0

No changes, remove nullsafety suffix

## 0.5.0-nullsafety.0

Add null safety

## 0.4.1

Add Github Actions and more tests

## 0.4.0

Upgrade diffutil.dart dependency to 1.0.0

## 0.3.0+1

Documentation fixes

## 0.3.0

Add DiffUtilSliverList.fromKeyedWidgetList! You can now implicitly a animate changes in a List of
Widgets, if every Widget has a unique key.

## 0.2.1

Fix generic type issue

## 0.2.0+3

- Improve sample

## 0.2.0+2

- Fix typo in documentation

## 0.2.0+1

- Documentation fixes

## 0.2.0

- *Breaking Change*: Remove Support for custom list-like types, made the API too complex and added too little benefit.
If there's need I will add specialized Widgets for custom list types.

- Fix bug where sometimes a wrong item was animated.

- Update example

## 0.1.1

Fix example

## 0.1.0

Depend on latest version of diffutil

## 0.0.2

Add more metadata

## 0.0.1

Preview version


test/diffutil_test.dart
import 'package:diffutil_sliverlist/diffutil_sliverlist.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  testWidgets("does not allow non-unique keys when using fromKeyedWidgetList",
      (tester) async {
    await tester.pumpWidget(
      Builder(
        builder: (context) => DiffUtilSliverList.fromKeyedWidgetList(
          children: [
            Container(
              key: const Key('a'),
            ),
            Container(
              key: const Key('a'),
            ),
          ],
          insertAnimationBuilder: (context, animation, widget) => widget,
          removeAnimationBuilder: (context, animation, widget) => widget,
        ),
      ),
    );

    final dynamic error = tester.takeException();

    expect(error, isAssertionError);
  });

  testWidgets("does not throw error when keys are unique", (tester) async {
    await tester.pumpWidget(
      MaterialApp(
        home: CustomScrollView(
          slivers: [
            DiffUtilSliverList.fromKeyedWidgetList(
              children: [
                Container(
                  key: const Key('a'),
                ),
                Container(
                  key: const Key('b'),
                ),
              ],
              insertAnimationBuilder: (context, animation, widget) => widget,
              removeAnimationBuilder: (context, animation, widget) => widget,
            ),
          ],
        ),
      ),
    );

    final dynamic error = tester.takeException();

    expect(error, isNull);
  });

  testWidgets("insert builder called", (tester) async {
    var insertCount = 0;

    await tester.pumpWidget(
      MaterialApp(
        home: CustomScrollView(
          slivers: [
            DiffUtilSliverList.fromKeyedWidgetList(
              children: [
                Container(
                  key: const Key('a'),
                ),
              ],
              insertAnimationBuilder: (context, animation, widget) => widget,
              removeAnimationBuilder: (context, animation, widget) => widget,
            ),
          ],
        ),
      ),
    );

    await tester.pumpWidget(
      MaterialApp(
        home: CustomScrollView(
          slivers: [
            DiffUtilSliverList.fromKeyedWidgetList(
              children: [
                Container(
                  key: const Key('a'),
                ),
                Container(
                  key: const Key('b'),
                ),
              ],
              insertAnimationBuilder: (context, animation, widget) {
                if (animation.isCompleted) {
                  expect(widget.key, const Key('a'));
                } else {
                  expect(widget.key, const Key('b'));
                }
                insertCount++;
                return widget;
              },
              removeAnimationBuilder: (context, animation, widget) {
                fail("removeAnimationBuilder should not be called");
              },
            ),
          ],
        ),
      ),
    );

    await tester.pumpAndSettle();

    expect(insertCount, 2);
  });

  testWidgets("remove builder called", (tester) async {
    var removeCount = 0;

    await tester.pumpWidget(
      MaterialApp(
        home: CustomScrollView(
          slivers: [
            DiffUtilSliverList.fromKeyedWidgetList(
              children: [
                Container(
                  key: const Key('a'),
                ),
                Container(
                  key: const Key('b'),
                ),
              ],
              insertAnimationBuilder: (context, animation, widget) => widget,
              removeAnimationBuilder: (context, animation, widget) => widget,
            ),
          ],
        ),
      ),
    );

    await tester.pumpWidget(
      MaterialApp(
        home: CustomScrollView(
          slivers: [
            DiffUtilSliverList.fromKeyedWidgetList(
              children: [
                Container(
                  key: const Key('b'),
                ),
              ],
              insertAnimationBuilder: (context, animation, widget) {
                expect(animation.isCompleted, isTrue);
                return widget;
              },
              removeAnimationBuilder: (context, animation, widget) {
                expect(widget.key, const Key('a'));
                expect(animation.isCompleted, isFalse);
                removeCount++;
                return widget;
              },
            ),
          ],
        ),
      ),
    );

    expect(removeCount, 1);
  });

  testWidgets("mutable list", (tester) async {
    final myList = ["1", "2", "3"];
    final rebuilder = ValueNotifier(0);

    int removeCount = 0;

    final widget = MaterialApp(
      home: CustomScrollView(
        slivers: [
          ValueListenableBuilder(
            valueListenable: rebuilder,
            builder: (BuildContext context, value, Widget? child) =>
                DiffUtilSliverList<String>(
                    items: myList,
                    builder: (context, item) => Text(item),
                    insertAnimationBuilder: (context, animation, widget) =>
                        SizeTransition(sizeFactor: animation, child: widget),
                    removeAnimationBuilder: (context, animation, widget) {
                      removeCount++;
                      return SizeTransition(
                          sizeFactor: animation, child: widget);
                    }),
          ),
        ],
      ),
    );
    await tester.pumpWidget(widget);

    await tester.pumpAndSettle();

    myList.removeLast();
    //trigger a rebuild manually
    rebuilder.value = rebuilder.value + 1;

    await tester.pumpAndSettle();

    expect(removeCount, 1);
  });
}


lib/src/diffutil_sliverlist_widget.dart
import 'package:diffutil_dart/diffutil.dart' as diffutil;
import 'package:flutter/foundation.dart';
import 'package:flutter/widgets.dart';

typedef EqualityChecker<T> = bool Function(T, T);
typedef DiffUtilWidgetBuilder<T> = Widget Function(BuildContext, T);
typedef AnimatedDiffUtilWidgetBuilder = Widget Function(
  BuildContext,
  Animation<double>,
  Widget,
);

class DiffUtilSliverList<T> extends StatefulWidget {
  /// the list of items for which changes are animated
  final List<T> items;

  /// builder that renders a single item (without animation)
  final Widget Function(BuildContext, T) builder;

  /// builder that renders the insertion animation
  final AnimatedDiffUtilWidgetBuilder insertAnimationBuilder;

  /// that renders the removal animation
  final AnimatedDiffUtilWidgetBuilder removeAnimationBuilder;
  final Duration insertAnimationDuration;
  final Duration removeAnimationDuration;

  final EqualityChecker<T>? equalityChecker;

  /// @param items a list of items to construct widgets from. Must implement == correctly if no equalityChecker is set.
  /// @param builder builds a widget from a given item
  /// @param insertAnimationBuilder The animation builder for insert animations
  /// @param removeAnimationBuilder The animation builder for insert animations
  /// @param insertAnimationDuration The duration of the insert animation
  /// @param removeAnimationDuration The duration of the remove animation
  /// @param equalityChecker optional custom equality implementation, defaults to ==
  const DiffUtilSliverList({
    Key? key,
    required this.items,
    required this.builder,
    required this.insertAnimationBuilder,
    required this.removeAnimationBuilder,
    this.insertAnimationDuration = const Duration(milliseconds: 300),
    this.removeAnimationDuration = const Duration(milliseconds: 300),
    this.equalityChecker,
  }) : super(key: key);

  /// Construct a animated list from a list widgets with unique keys.
  ///
  /// @param children A List a Widgets with unique keys
  /// @param insertAnimationBuilder The animation builder for insert animations
  /// @param removeAnimationBuilder The animation builder for insert animations
  /// @param insertAnimationDuration The duration of the insert animation
  /// @param removeAnimationDuration The duration of the remove animation
  static DiffUtilSliverList<Widget> fromKeyedWidgetList({
    required List<Widget> children,
    required AnimatedDiffUtilWidgetBuilder insertAnimationBuilder,
    required AnimatedDiffUtilWidgetBuilder removeAnimationBuilder,
    Duration insertAnimationDuration = const Duration(milliseconds: 300),
    Duration removeAnimationDuration = const Duration(milliseconds: 300),
  }) {
    if (kDebugMode) {
      final keys = <Key?>{};
      for (final child in children) {
        if (!keys.add(child.key)) {
          throw FlutterError(
            'DiffUtilSliverList.fromKeyedWidgetList called with widgets that do not contain unique keys! '
            'This is an error as changed is this list cannot be animated reliably. Use unique keys or the default constructor. '
            'This duplicate key was ${child.key} in widget $child.',
          );
        }
      }
    }
    return DiffUtilSliverList<Widget>(
      items: children,
      builder: (context, widget) => widget,
      insertAnimationBuilder: insertAnimationBuilder,
      removeAnimationBuilder: removeAnimationBuilder,
      insertAnimationDuration: insertAnimationDuration,
      removeAnimationDuration: removeAnimationDuration,
      equalityChecker: (a, b) => a.key == b.key,
    );
  }

  @override
  _DiffUtilSliverListState<T> createState() => _DiffUtilSliverListState<T>();
}

class _DiffUtilSliverListState<T> extends State<DiffUtilSliverList<T>> {
  late GlobalKey<SliverAnimatedListState> listKey;

  late List<T> oldList;

  @override
  void initState() {
    super.initState();
    listKey = GlobalKey<SliverAnimatedListState>();
    oldList = List.from(widget.items);
  }

  @override
  void didUpdateWidget(DiffUtilSliverList<T> oldWidget) {
    super.didUpdateWidget(oldWidget);
    final newList = widget.items;

    final diff = diffutil
        .calculateListDiff<T>(
          oldList,
          newList,
          detectMoves: false,
          equalityChecker: widget.equalityChecker,
        )
        .getUpdates();
    //tmp copy of old to start the animations on the right positions
    final tmpList = List<T?>.from(oldList);
    for (final update in diff) {
      _onDiffUpdate(update, tmpList);
    }
    oldList = List.from(newList);
  }

  @override
  Widget build(BuildContext context) {
    return SliverAnimatedList(
      key: listKey,
      initialItemCount: widget.items.length,
      itemBuilder: (context, index, animation) => widget.insertAnimationBuilder(
        context,
        animation,
        widget.builder(context, widget.items[index]),
      ),
    );
  }

  void _onChanged(int position, Object? payload, final List<T?> tmpList) {
    listKey.currentState!.removeItem(
      position,
      (context, animation) => const SizedBox.shrink(),
      duration: Duration.zero,
    );
    _onInserted(position, 1, tmpList);
  }

  void _onInserted(
    final int position,
    final int count,
    final List<T?> tmpList,
  ) {
    for (var loopCount = 0; loopCount < count; loopCount++) {
      listKey.currentState!.insertItem(
        position + loopCount,
        duration: widget.insertAnimationDuration,
      );
    }
    tmpList.insertAll(position, List<T?>.filled(count, null));
  }

  void _onRemoved(final int position, final int count, final List<T?> tmpList) {
    for (var loopCount = 0; loopCount < count; loopCount++) {
      // ignore: null_check_on_nullable_type_parameter
      final oldItem = tmpList[position + loopCount]!;
      // i purposefully remove the item at the same position on each
      // turn. the internal state is updated, so it removes the right item
      // actually. i only need to calculate the position of oldList
      // which might get ot of sync if count > 1.
      // the tempList is only updated at the end of the method for better performance
      listKey.currentState!.removeItem(
        position,
        (context, animation) => widget.removeAnimationBuilder(
          context,
          animation,
          widget.builder(context, oldItem),
        ),
        duration: widget.removeAnimationDuration,
      );
    }
    tmpList.removeRange(position, position + count);
  }

  void _onDiffUpdate(diffutil.DiffUpdate update, List<T?> tmpList) {
    update.when<void>(
      move: (_, __) =>
          throw UnimplementedError('moves are currently not supported'),
      insert: (pos, count) => _onInserted(pos, count, tmpList),
      change: (pos, payload) => _onChanged(pos, payload, tmpList),
      remove: (pos, count) => _onRemoved(pos, count, tmpList),
    );
  }
}


lib/src/sliver_list_update_adapter.dart
import 'package:diffutil_dart/diffutil.dart' as diffutil;
import 'package:diffutil_sliverlist/src/diffutil_sliverlist_widget.dart';
import 'package:flutter/widgets.dart';

class ListUpdateCallBackToSliverAnimatedListKeyAdapter<T>
    implements diffutil.ListUpdateCallback {
  final GlobalKey<SliverAnimatedListState> stateKey;
  final AnimatedDiffUtilWidgetBuilder removeAnimationBuilder;
  final DiffUtilWidgetBuilder<T> builder;
  final List<T> oldList;
  final Duration insertDuration;
  final Duration removeDuration;

  ListUpdateCallBackToSliverAnimatedListKeyAdapter(
      this.stateKey,
      this.builder,
      this.oldList,
      this.removeAnimationBuilder,
      this.insertDuration,
      this.removeDuration);

  @override
  void onChanged(int position, int count, Object payload) {
    while (count > 0) {
      stateKey.currentState.removeItem(
          position, (context, animation) => const SizedBox.shrink(),
          duration: const Duration());
      onInserted(position, 1);
      count--;
      position++;
    }
  }

  @override
  void onInserted(final int position, final int count) {
    for (var loopCount = 0; loopCount < count; loopCount++) {
      stateKey.currentState
          .insertItem(position + loopCount, duration: insertDuration);
    }
    oldList.insertAll(position, List<T>.filled(count, null));
  }

  @override
  void onMoved(int fromPosition, int toPosition) {
    /// [SliverAnimatedList] does not support moves
    throw UnimplementedError();
  }

  @override
  void onRemoved(final int position, final int count) {
    for (var loopCount = 0; loopCount < count; loopCount++) {
      final oldItem = oldList[position + loopCount];
      // i purposfully remove the item at the same position on each
      // turn. the internal state is updated, so it removes the right item
      // actually. i only need to calculate the positon of oldLost
      // which might get ot of sync if count > 1.
      // the oldList is only updated at the end of the method for better performance
      stateKey.currentState.removeItem(
          position,
          (context, animation) => removeAnimationBuilder(
              context, animation, builder(context, oldItem)),
          duration: removeDuration);
    }
    oldList.removeRange(position, position + count);
  }
}


example/lib/main.dart
import 'dart:math';

import 'package:diffutil_sliverlist/diffutil_sliverlist.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

void main() => runApp(const DiffUtilSliverListDemo());

class DiffUtilSliverListDemo extends StatefulWidget {
  const DiffUtilSliverListDemo({super.key});

  @override
  _DiffUtilSliverListDemoState createState() => _DiffUtilSliverListDemoState();
}

class _DiffUtilSliverListDemoState extends State<DiffUtilSliverListDemo> {
  int _counter = 0;

  List<int> list = [0];

  void _incrementCounter() {
    setState(() {
      _counter++;

      if (Random().nextInt(3) > 0 || list.isEmpty) {
        list.add(_counter);
      } else {
        list.removeAt(Random().nextInt(list.length));
      }
    });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: DefaultTabController(
        length: 2,
        initialIndex: 0,
        child: Scaffold(
          appBar: AppBar(
            bottom:
                TabBar(tabs: [Tab(text: "example 1"), Tab(text: "example 2")]),
          ),
          body: TabBarView(
            children: [
              Scaffold(
                  floatingActionButton: FloatingActionButton(
                    onPressed: _incrementCounter,
                    tooltip: 'Increment',
                    child: Icon(Icons.view_list),
                  ), // This tr
                  body: CustomScrollView(
                    slivers: [
                      DiffUtilSliverList<int>(
                        items: list,
                        builder: (context, item) => Container(
                          color: colors[item % colors.length],
                          height: 48,
                          width: double.infinity,
                        ),
                        insertAnimationBuilder: (context, animation, child) =>
                            FadeTransition(
                          opacity: animation,
                          child: child,
                        ),
                        removeAnimationBuilder: (context, animation, child) =>
                            SizeTransition(
                          sizeFactor: animation,
                          child: child,
                        ),
                        removeAnimationDuration:
                            const Duration(milliseconds: 300),
                        insertAnimationDuration:
                            const Duration(milliseconds: 120),
                      ),
                    ],
                  )),
              CustomScrollView(
                slivers: <Widget>[
                  SliverToBoxAdapter(
                    child: Padding(
                        padding: const EdgeInsets.all(16.0),
                        child: Text(
                          "FIRST LIST",
                          style: Theme.of(context).textTheme.subtitle2,
                        )),
                  ),
                  ExpandableLists(),
                  SliverToBoxAdapter(
                    child: Padding(
                      padding: const EdgeInsets.all(16.0),
                      child: Text(
                        "SECOND LIST",
                        style: Theme.of(context).textTheme.subtitle2,
                      ),
                    ),
                  ),
                  ExpandableLists(),
                ],
              ),
            ],
          ),
        ),
      ),
    );
  }

  static const colors = [
    Colors.deepOrangeAccent,
    Colors.blueAccent,
    Colors.greenAccent,
    Colors.yellowAccent
  ];
}

class ExpandableLists extends StatefulWidget {
  @override
  _ExpandableListsState createState() => _ExpandableListsState();
}

class _ExpandableListsState extends State<ExpandableLists> {
  bool expanded = false;

  @override
  Widget build(BuildContext context) {
    return DiffUtilSliverList.fromKeyedWidgetList(
      children: [
        ListTile(
          key: Key("1"),
          title: Text("first"),
          trailing: Icon(Icons.chevron_right),
        ),
        ListTile(
          key: Key("2"),
          title: Text("second"),
          trailing: Icon(Icons.chevron_right),
        ),
        if (this.expanded)
          for (int i = 3; i < 6; i++)
            ListTile(
              key: Key(i.toString()),
              title: Text("index: $i"),
              trailing: Icon(Icons.chevron_right),
            ),
        ListTile(
          key: Key("expand_collapse"),
          onTap: () => setState(() {
            expanded = !expanded;
          }),
          title: Text(
            expanded ? "collapse" : "expand",
            style: TextStyle(fontWeight: FontWeight.bold),
          ),
          trailing: Icon(expanded ? Icons.expand_less : Icons.expand_more),
        )
      ],
      insertAnimationBuilder: (context, animation, child) => SizeTransition(
          sizeFactor: animation.drive(Tween<double>(begin: 0.7, end: 1)),
          axisAlignment: 0,
          child: FadeTransition(
            opacity: animation,
            child: child,
          )),
      removeAnimationBuilder: (context, animation, child) => FadeTransition(
        opacity: animation,
        child: SizeTransition(
          sizeFactor: animation,
          axisAlignment: 0,
          child: child,
        ),
      ),
    );
  }
}
